### ISSUE TITLE
Document 11LOGO

### ISSUE THREAD

**Issue by larsbrinkhoff** [PDP-10/its#1192]
@Alexey-Slyusar is working on this:

https://drive.google.com/file/d/1gKXYHLcV7MyXINfvZeth4wSxq32_42Nc/view?usp=sharing

**Comment by larsbrinkhoff** [comment]
Some notes:
- KILLDISPLAY is probably the opposite of STARTDISPLAY.
- I wonder how to use a display other than the upper left?  Maybe that's not supported in this version.
- The turtle should be a triangle, but isn't.  Also, I think LEFT doesn't always turn the right angle.  There may be bugs.
- RUG probably jumps to the PDP-11 debugger.  We have it (RUG; AR >), but how to get both that and 11LOGO running we don't know yet.
- VERSION, 1005, is the sum of all source file versions.
- What is THING?
- DO can evaluate a list.
- Are there any line editing characters?  Backspace doesn't work, is there another way to delete?

**Comment by Alexey-Slyusar** [comment]
Here are comments:
Some notes:
- KILLDISPLAY is probably the opposite of STARTDISPLAY.
  Added
- I wonder how to use a display other than the upper left?  Maybe that's not supported in this version.
  It seems that it is so :(
- The turtle should be a triangle, but isn't.  Also, I think LEFT doesn't always turn the right angle.  There may be bugs.
  Yes that's right. Here is an example shows the bug:
  SQUARES procedure with RIGHT
![img_4877](https://user-images.githubusercontent.com/42380964/44933916-99eb9e80-ad73-11e8-817b-871b81a762b8.jpg)
   The same SQUARE procedure but with LEFT
![img_4871](https://user-images.githubusercontent.com/42380964/44933956-cc959700-ad73-11e8-9ce3-61af9091139b.jpg)

- RUG probably jumps to the PDP-11 debugger.  We have it (RUG; AR >), but how to get both that and 11LOGO running we don't know yet.
   Added
- VERSION, 1005, is the sum of all source file versions.
   There is VERSION primitive in 11LOGO that outputs the version number. Am I wrong?
- What is THING?
   Added
- DO can evaluate a list.
   Added.
- Are there any line editing characters?  Backspace doesn't work, is there another way to delete?
  Both AI MEMO 315 and AI MEMO 315a say that rubout (Del key) should delete previous character but this doesn't work as Ctrl + W and other editing characters either.

**Comment by Alexey-Slyusar** [comment]
Here is updated version of the document:

[A GLOSSARY OF 11LOGO PRIMITIVES_v0.2](https://drive.google.com/file/d/1bQApHW1QBj8xPJJeIJfTUMWjVhUnSbbN/view?usp=sharing)

**Comment by larsbrinkhoff** [comment]
- I made a separate issue for the LEFT bug.
- Right, the VERSION primitive prints the version which is the same number printed when 11LOGO starts.

**Comment by larsbrinkhoff** [comment]
If you can work with git, I suggest making a new branch e.g. alexey/11logo to add your document.  Maybe doc/_info_/11logo.order.

**Comment by Alexey-Slyusar** [comment]
@larsbrinkhoff, I want to check the 11logo.order against 11LOGO system file
https://github.com/PDP-10/its-vault/tree/master/files/11logo
but it seems that the link is not actual anymore.
Could you please provide an actual link to the 11LOGO system file?

**Comment by Alexey-Slyusar** [comment]
It seems that there is no THING built-in procedure in our 11LOGO version. There is only dots (:) as an abbreviation for the THING-qote (e.g. THING "VARIABLE_NAME) so:

?MAKE "N 100
?PRINT THING "N
047372 ... HAS NO MEANING.

but short form works:

?PRINT :N 
100

SHOW lists a procedure, but the name of the procedure must be quoted:
?SHOW "CS

TO CS
10 WIPECLEAN
20 SETTURTLE HOME
END

And that name for such procedure is very unusual both for early and for modern logo versions. So I had to do some more try and error iterations to figure out what the SHOW do in 11LOGO 1005 :-)

It seems that there is no MODE primitive in our 11LOGO version.

**Comment by larsbrinkhoff** [comment]
11LOGO; SYSTEM 327 is here:  
https://github.com/PDP-10/its-vault/blob/master/files/11logo/system.327

The command table starts at line 863.

**Comment by Alexey-Slyusar** [comment]
Thank you!
Some moments to discuss:
* The 11LOGO; SYSTEM 327 has a description for AND noise word. But the interpreter complains "HAS NO MEANING" for the AND. So SUM 3 AND 4 doesn't work. At the same time another noise word OF seems to be acceptable by the interpreter so SUM OF 3 4 works well.
* It seems that there are descriptions for editing characters in the 11LOGO; SYSTEM 327 . In particular for Ctrl + W (in accordance with AIM 315 it deletes previous word).
1003 ;^W	FSF .   .   .    .   NNUMF .    	20
But despite that, editing characters don't work.
* I am still don't get MOD primitive meaning.

**Comment by Alexey-Slyusar** [comment]
I've found that the MOD procedure accepts two inputs. Both must be numbers.
But I am still trying to figure out the purpose of the MOD procedure.
Here are some examples:
?MOD 1 12
016560 WHAT SHOULD I DO WITH 0
?MOD -14 11
016560 WHAT SHOULD I DO WITH 0
?MOD 1 0
044110 ARITHMETIC RESULT TOO BIG.
?MOD 1 0.1
047372 2A. HAS NO MEANING.
?MOD -14 -14
047260 UNEXPECTED END OF LINE.
?MOD -5 0
044110 ARITHMETIC RESULT TOO BIG.

**Comment by larsbrinkhoff** [comment]
It's related to division.  If we look into EVAL 336, we can see MOD clears the D register and then jumps to DIV.1.  It looks like the intent is to compute the modulo function.

**Comment by Alexey-Slyusar** [comment]
Thank you, Lars! 

If I understood correctly, then it looks like MOD procedure represents modulo function and evaluating the expression ?PRINT MOD 21 4 should result 1 (In fact CLOGO has MOD procedure and it works like that, later Logo versions, including 11LOGO described in AIM 315, have REMAINDER procedure with the same meaning). But in our 11LOGO version it looks like this:
?PRINT MOD 21 4
0
?
So it seems that the MOD is buggy or has a different meaning.

Besides this, I basically finished the document.

**Comment by larsbrinkhoff** [comment]
The implementation of MOD starts here: https://github.com/PDP-10/its/blob/master/src/11logo/eval.336#L503

There's a trick to jump to different places depending on if MOD or DIVDE is called: https://github.com/PDP-10/its/blob/master/src/11logo/eval.336#L515  When MOD is called D is set to 0, so the `ADD D,PC` does not jump and we get to `DIV.8` which looks like it should get the modulo result.

I think there's just a bug somewhere.  It could be in 11LOGO or 11SIM.

**Comment by larsbrinkhoff** [comment]
We should be able to get 11LOGO running in e.g. SIMH.  Just without with turtle graphics display.  If MOD behaves the same, we can conclude the bug is probably in 11LOGO.

To do this, write 11LOGO BIN to a tape.  See https://github.com/PDP-10/its/blob/master/doc/DUMP-itstar.md for how.  Then translate the BIN file to PDP-11 absolute loader format with tools/dasm/palx.  The resulting file should load into SIMH with the `load` command.

**Comment by Alexey-Slyusar** [comment]
Thanks for the explanation!
If I understand correctly, the idea is to get 11LOGO running in SIMH PDP-11
simulation and check its behavior. It is very interesting.
I going to check the things. But it takes time since I don't familiar with
SIMH PDP-11 simulation.

ср, 5 сент. 2018 г. в 9:57, Lars Brinkhoff <notifications@github.com>:

**Comment by Alexey-Slyusar** [comment]
I've wrote 11LOGO to a tape. 
But there is no prebuilt executable called palx in my tools/dasm/ directory.
So I've found the palx.c source file in the git master brunch and tried to build it but unsuccesfully:

~/tools/gcc palx.c 
/tmp/cckKZWWG.o: In function `pdp11_in':
palx.c:(.text+0x47): undefined reference to `get_word'
palx.c:(.text+0x57): undefined reference to `get_word'
/tmp/cckKZWWG.o: In function `absolute_out':
palx.c:(.text+0x138): undefined reference to `get_word'
palx.c:(.text+0x167): undefined reference to `get_word'
/tmp/cckKZWWG.o: In function `image_out':
palx.c:(.text+0x1fc): undefined reference to `get_word'
/tmp/cckKZWWG.o:palx.c:(.text+0x23a): more undefined references to `get_word' follow
/tmp/cckKZWWG.o: In function `main':
palx.c:(.text+0x2c2): undefined reference to `file_36bit_format'
palx.c:(.text+0x2eb): undefined reference to `parse_word_format'
collect2: error: ld returned 1 exit status

@larsbrinkhoff, could you please advise how can I compile palx or get it in binary format? 
Thank you!

**Comment by larsbrinkhoff** [comment]
I believe you will have to cd to tools/dasm and update that sub repository from GitHub.  Something like this:

```
cd tools/dasm
git fetch
git merge --ff-only origin/master
make
```

palx.c needs to be linked with some other files, so just `gcc palx.c` doesn't work.

**Comment by Alexey-Slyusar** [comment]
Lars, thank you!
I've got palx :)

But I cannot translate the l.bin file to PDP-11 absolute loader format, because I can't fugure out how to use palx properly:

.../its$ tools/dasm/./palx -A l.bin 
Usage: optind -I|A [-W<format>]

Give me a hint, please :-)
Thank you in advance and sorry for such silly question.

**Comment by larsbrinkhoff** [comment]
No problem at all.  Use redirection: palx < input > output.

**Comment by Alexey-Slyusar** [comment]
Thank you!
I've made one step futher.
/its$ ./palx -A <l.bin> 11l.bin
/its$ pdp11

PDP-11 simulator V3.8-1
sim> load 11l.bin
sim> 

But I am not sure that am I on right track and what should I do next to get 11LOGO up and running on SIMH PDP11 ...

**Comment by larsbrinkhoff** [comment]
It's good SIMH accepted the `load` command.  This means your 11l.bin file should be good.

Next, type `run 400` to start.

**Comment by Alexey-Slyusar** [comment]
Lars,
an error occurred while loading the 11l.bin into SIMH PDP-11. Here is a screenshot:

![default](https://user-images.githubusercontent.com/42380964/45237611-df661980-b2e7-11e8-84d7-0390ba1140d0.png)

If I am not mistaken, I got the same error as you described in 11LOGO #889.
You wrote that: "Attaching some needed devices did the trick!". 
Do you have any ideas how to solve the current issue?

Thank you!

**Comment by larsbrinkhoff** [comment]
These SIMH commands should match the 11SIM setup:

```
set cpu 11/45
set cpu 56k
set rf enabled
at rf0 rf0.dsk
set pclk enabled

**Comment by Alexey-Slyusar** [comment]
the set cpu 56k command was not accepted by the simulator:
![default](https://user-images.githubusercontent.com/42380964/45239246-43d7a780-b2ed-11e8-86c4-6d76ef18379d.png)

**Comment by larsbrinkhoff** [comment]
Strange, my copy of SIMH does accept `set cpu 56k`.  Maybe try the latest version from GitHub?

Looks like HK conflicts with RF, so `set hk disabled`.

**Comment by Alexey-Slyusar** [comment]
I tried latest prebuilt version from github, but unsuccessfully:

PDP-11 simulator V4.0-0 Current        git commit id: b007669a
sim> set cpu 11/45
Disabling XQ
sim> set cpu 56k
Non-existent parameter
sim> set rf enabled
sim> at rf0 rf0.dsk
RF: creating new file
RF: buffering file in memory
sim> set hk disabled
sim> set pclk enabled
sim> set cpu 56k
Non-existent parameter
sim> load 11l.bin
sim> run 400
I GIVE UP!VECTOR 0 TRAPPED FROM 0
HALT instruction, PC: 006762 (BR 6760)
sim>

I am going to compile simh pdp11 from source code from the github repository's master branch

**Comment by Alexey-Slyusar** [comment]
I've got pdp11 compiled from the github repository's master branch. And the set cpu 56k command works fine in the latest pdp11 version. Thank you!
But there is still an error:
![default](https://user-images.githubusercontent.com/42380964/45241650-e09e4300-b2f5-11e8-9b86-c9218072643e.png)

**Comment by larsbrinkhoff** [comment]
I found the problem.  You also need to `set ke enabled`.  I used this with success:

```
set cpu 11/45
set cpu 56k
set hk disabled
set rf enabled
at rf0 rf0.dsk
set pclk enabled
set ke enabled
load 11logo.lda
go 400

**Comment by Alexey-Slyusar** [comment]
Great!

Unfortunately 11LOGO MOD has the same behaviour as in PDP45 under ITS: 
![default](https://user-images.githubusercontent.com/42380964/45256974-eb64e080-b3a6-11e8-9f03-8af30d331561.png)

So, the bad news is that our 11LOGO version has a bug (or probably bugs).
But the good news is that the PDP45 works fine :-)

I've updated the 11logo.order with the result of our exploration. I think that the document is finished.

**Comment by Alexey-Slyusar** [comment]
Actually, I've checked a few more things than just MOD procedure (without turtle graphics of course). 
AND and the editing characters (like del key, ^W, ^Y etc.) do not work under SIMH PDP11 simulation either. But other things work the same way. So it seems that we have exactly the same 11LOGO behavior in SIMH PDP11 in comparison with PDP45 simulation on ITS.

**Comment by larsbrinkhoff** [comment]
Very good.  So we can be pretty sure the bugs are in 11LOGO rather than 11SIM.

Please make pull requests for 11LOGO and CLOGO documention when you feel ready.

**Comment by larsbrinkhoff** [comment]
11LOGO done: #1217.

**Comment by larsbrinkhoff** [comment]
Fixed by #1217.
